### PR TITLE
added component type backing form validator

### PR DESCRIPTION
**Pull request by debonair**
@danfuterman please review

### CODE REVIEW

**Comment by danfuterman** [comment]
@bausmeier could you do an initial review of this and the related frontend PR -https://github.com/jembi/bsis-frontend/pull/162.

File: src/backingform/validator/ComponentTypeBackingFormValidator.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.Arrays;
+
+public class ComponentTypeBackingFormValidator  implements Validator {
+
+    private Validator validator;
+    private UtilController utilController;
+    private ProductTypeRepository productTypeRepository;
+
+    public ComponentTypeBackingFormValidator(Validator validator, UtilController utilController, ProductTypeRepository productTypeRepository){
```
Comment (bausmeier) [inline src/backingform/validator/ComponentTypeBackingFormValidator.java]: `utilController` and `productTypeRepository` are unused and should be removed.

File: src/backingform/validator/ComponentTypeBackingFormValidator.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    public boolean supports(Class<?> clazz) {
+        return Arrays.asList(ComponentTypeBackingForm.class,
+                ProductTypeRepository.class, String.class).contains(clazz);
+    }
+
+    @Override
+    public void validate(Object obj, Errors errors){
+
+        ValidationUtils.invokeValidator(validator, obj, errors);
+        ComponentTypeBackingForm form = (ComponentTypeBackingForm) obj;
```
Comment (bausmeier) [inline src/backingform/validator/ComponentTypeBackingFormValidator.java]: `form` is unused.

File: src/backingform/validator/ComponentTypeBackingFormValidator.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        super();
+        this.validator = validator;
+        this.utilController = utilController;
+        this.productTypeRepository = productTypeRepository;
+    }
+
+    @Override
+    public boolean supports(Class<?> clazz) {
+        return Arrays.asList(ComponentTypeBackingForm.class,
+                ProductTypeRepository.class, String.class).contains(clazz);
```
Comment (bausmeier) [inline src/backingform/validator/ComponentTypeBackingFormValidator.java]: Why does this validator support `ProductTypeRepository` and `String`?

**Comment by bausmeier** [comment]
@debonair Please merge master into this branch as resolve the merge conflicts.

**Comment by bausmeier** [comment]
If I try to create a component type via the POST endpoint I get a NullPointerException. This is caused by `hasBloodGroup` or `producedComponentTypeCombinations` not being provided. So either these need to be included in the validation, or some null checks need to be added before accessing them in the `ComponentTypeViewModel`. See stack traces below.

```
Caused by: java.lang.NullPointerException
	at viewmodel.ComponentTypeViewModel.getHasBloodGroup(ComponentTypeViewModel.java:42)
```

```
Caused by: java.lang.NullPointerException
	at viewmodel.ComponentTypeViewModel.getComponentTypeCombinationViewModels(ComponentTypeViewModel.java:87)
	at viewmodel.ComponentTypeViewModel.getProducedComponentTypeCombinations(ComponentTypeViewModel.java:78)
```

File: src/backingform/validator/ComponentTypeBackingFormValidator.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.validation.ValidationUtils;
+import org.springframework.validation.Validator;
+
+import java.util.Arrays;
+
+public class ComponentTypeBackingFormValidator implements Validator {
+
+    private Validator validator;
+
+    public ComponentTypeBackingFormValidator(Validator validator) {
```
Comment (bausmeier) [inline src/backingform/validator/ComponentTypeBackingFormValidator.java]: Sorry I missed this before, but you don't need to pass in the validator here. The validation works without it so you can remove it and the call to `invokeValidator` in the validate method below. So you can actually take out this whole constructor and just have the methods overriding supports and validate.

**Comment by bausmeier** [comment]
This validator is a good candidate for some unit tests since it doesn't require any database access.

@danfuterman Is it worth writing the tests now or should they be done at a later point?

**Comment by danfuterman** [comment]
@bausmeier - I think we need to document an approach (or at least have a few examples to reference) for adding validator unit tests before asking @debonair to add them in here. If you have the time, it would be great if you added a few unit tests for this that could form the reference-able example for next time, but it's not expected of you.

**Comment by bausmeier** [comment]
@danfuterman @debonair I added tests for the validator. See https://github.com/jembi/bsis/commit/461ee924e9f9b70c581747399db21ea5b77c868b.

**Comment by danfuterman** [comment]
Thanks @bausmeier looks great

**Comment by bausmeier** [comment]
@debonair @danfuterman What is happening with this PR? I see that it is assigned to me.

**Comment by danfuterman** [comment]
@bausmeier @debonair re-assigned to myself, there are a few edits/updates that I need to make to this.

**Pull request closed without merging**